src/wasm-type.h

```diff
@@ -449,6 +455,42 @@ struct Rtt {
   std::string toString() const;
 };
 
+// TypeBuilder - allows for the construction of recursive types. Contains a
+// table of `n` mutable HeapTypes and can construct temporary types that are
+// backed by those HeapTypes, refering to them by reference. Those temporary
+// types are owned by the TypeBuilder and should only be used in the
+// construction of HeapTypes to insert into the TypeBuilder. Temporary types
+// should never be used in the construction of normal Types, only other
+// temporary types.
+struct TypeBuilder {
+  struct Impl;
+  std::unique_ptr<Impl> impl;
```

Comment:
We need to hide the definition of `Impl` in the .cpp file because it depends on implementation details like `HeapTypeInfo`. We can't do that and have `Impl impl;` because hiding the definition necessarily makes `Impl` and incomplete type at this point. This is the classic [`PImpl` idiom](https://en.cppreference.com/w/cpp/language/pimpl).